frontends/p4/entryPriorities.cpp

```diff
@@ -0,0 +1,160 @@
+
+    size_t currentPriority = 1;
+    if (withPriority == nullptr) {
+        // no priorities specified
+        for (size_t i = 0; i < entries->size(); ++i) {
+            size_t index = largestWins ? entries->size() - i - 1 : i;
+            auto entry = entries->entries.at(index);
+            BUG_CHECK(entry->priority == nullptr, "%1%: Priority found?", entry);
+            auto priority = new IR::Constant(currentPriority);
+            auto newEntry = new IR::Entry(entry->srcInfo, entry->annotations, entry->isConst,
+                                          priority, entry->keys, entry->action, entry->singleton);
+            entries->entries[index] = newEntry;
+            size_t nextPriority = currentPriority + priorityDelta;
+            if (nextPriority < currentPriority) {
+                ::error(ErrorType::ERR_OVERLIMIT, "%1% Overflow in priority computation", table);
+                return entries;
+            }
+            currentPriority = nextPriority;
+        }
+        return entries;
+    }
+
+    // Some priorities specified.
+    bool requiresPriorities = false;
+    auto key = table->getKey();
+    for (auto element : key->keyElements) {
+        if (requiresPriority(element)) {
+            requiresPriorities = true;
+            break;
+        }
+    }
+    if (!requiresPriorities) {
+        ::error(ErrorType::ERR_UNSUPPORTED,
+                "%1% key match type does not require priorities, but some are specified", key);
+        return entries;
+    }
+
+    std::map<size_t, const IR::Entry *> usedPriorities;
+    size_t previousPriority = 0;
+    for (size_t i = 0; i < entries->size(); ++i) {
+        auto entry = entries->entries.at(i);
+        if (entry->priority) {
+            if (auto value = entry->priority->to<IR::Constant>()) {
+                currentPriority = value->asUnsigned();
+            } else {
+                ::error(ErrorType::ERR_INVALID, "%1% must be a constant", entry->priority);
+                return entries;
+            }
+        } else {
+            // First entry must have a priority.
+            if (i == 0) {
+                ::error(ErrorType::ERR_EXPECTED, "%1% entry must have a priority", entry);
+                return entries;
+            }
+            auto priority = new IR::Constant(currentPriority);
+            auto newEntry = new IR::Entry(entry->srcInfo, entry->annotations, entry->isConst,
+                                          priority, entry->keys, entry->action, entry->singleton);
+            entries->entries[i] = newEntry;
+        }
+        auto previous = usedPriorities.find(currentPriority);
+        if (previous != usedPriorities.end()) {
+            warn(ErrorType::WARN_DUPLICATE_PRIORITIES, "%1% and %2% have the same priority %3%",
+                 previous->second, entries->entries[i], currentPriority);
+        }
+        usedPriorities.emplace(currentPriority, entries->entries[i]);
+        if (i > 0 && currentPriority < previousPriority) {
```

Comment:
I think this condition needs to be something like this:

```
i > 0 && ((largestWins && currentPriority > previousPriority) || (!largestWins && currentPriority < previousPriority))
```
As written now, it produces warnings when largestWins is true, in cases where it should not.